File: 0202-happy-number/0202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = n;

        do {
            slow = square(slow);
            fast = square(square(fast));
        } while (slow != fast);

        if (slow == 1)
            return true;
        
        return false;
    }

    public int square(int number) {
        int result = 0, remainder;

        while (number > 0) {
            remainder = number % 10;
            result += (remainder * remainder);
            number /= 10;
        }
        
        return result;
    }
}

File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
      
        def sum_of_squares(num):
            result = 0

            while num > 0:
                remainder = num % 10
                result += remainder ** 2
                num //= 10

            return result

        slow = fast = n

        while True:
            slow = sum_of_squares(slow)
            fast = sum_of_squares(sum_of_squares(fast))

            if slow == fast:
                break

        return slow == 1
